### ISSUE TITLE
Is this a bug in the code or my misunderstanding?

### ISSUE THREAD

**Issue by l11x0m7** [RaRe-Technologies/gensim#1873]
In `gensim/gensim/models/word2vec.py`, line 394 and line 401

```
    if learn_vectors:
        # learn input -> hidden, here for all words in the window separately
        if is_ft:
            **if not model.cbow_mean and input_word_indices:**
                neu1e /= (len(input_word_indices[0]) + len(input_word_indices[1]))
            for i in input_word_indices[0]:
                context_vectors_vocab[i] += neu1e * context_locks_vocab[i]
            for i in input_word_indices[1]:
                context_vectors_ngrams[i] += neu1e * context_locks_ngrams[i]
        else:
            **if not model.cbow_mean and input_word_indices:**
                neu1e /= len(input_word_indices)
            for i in input_word_indices:
                context_vectors[i] += neu1e * context_locks[i]

    return neu1e
```

Shouldn't this be `if model.cbow_mean and input_word_indices` rather than `if not model.cbow_mean and input_word_indices`?

**Comment by menshikh-iv** [comment]
CC @manneshiva @jayantj @gojomo

**Comment by gojomo** [comment]
No. If you've already averaged all the vectors together (`cbow_mean` true mode), then the error on the (average-magnitude) averaged-vector can also be applied full-magnitude to the the constituent vectors. If instead you've summed them together (the rarer `cbow_mean` false mode), then the error on the (possibly-high-magnitude) summed vector must be divided equally among the constituent vectors. 

See also discussion/example at https://groups.google.com/forum/#!searchin/gensim/cbow_mean|sort:date/gensim/BtN7uB1vgpc/tlvkLXqzJwAJ and prior issue #697.

**Comment by piskvorky** [comment]
Worth a code comment if it keep tripping up people?

**Comment by gojomo** [comment]
Sure, both the Python and Cython lines where this conditional division is done could include a extra clarification that this is intentional and even an excerpt from, or link to, these longer explanations.

**Comment by l11x0m7** [comment]
Yeah, that will be great!

**Comment by piskvorky** [comment]
@gojomo @l11x0m7 can you open a PR?

**Comment by thegrimreaper1992** [comment]
I also think this is wrong and I'm not convinced by @gojomo 's logic. Backprop is basically the chain rule nothing more. So if neu1e is the derivative wrt l1 and l1 is an average of some weights, then derivative of neu1e wrt the weights would be 1/(number of words) so I also think it's a bug. If you want to convince me you have to convince me with the application of chain rule not the thing you mentioned because I guess what you said is also wrong

**Comment by gojomo** [comment]
@thegrimreaper1992 The current code works well, and matches the behavior (in both code & rough results) of the original word2vec.c implementation from the authors of the original Word2Vec paper. If neither my explanation nor that concordance convinces you, go ahead and try it your way, and see if anything improves.

**Comment by thegrimreaper1992** [comment]
Thanks for your response @gojomo.
Honestly, I'm mostly trying to understand if I'm being mistaken or this is just different than what is explained in the paper. It's not really about getting better results. Or if this is a slight modification but they haven't mentioned it I was wondering why it makes sense. 

So in other words maybe I'm trying to understand why the original code was implemented that way. Because by applying chain rule I guess it should be the other way around. Do you get my point here?

**Comment by gojomo** [comment]
Sure, but note it's not just simple backprop, but backprop through this algorithm-specific "input-prep-operation", that's composed (in this particular CBOW-average mode) by the average of all context-window vectors. So at this level, we've essentially left the standard NN model, from which we have an error-at-the-"input"-layer. But that input-layer was composed in a certain way from all the other word-vectors, and that error then needs to be meaningfully applied back to constituent word-vectors. 

I think my forum post – https://groups.google.com/forum/#!searchin/gensim/cbow_mean%7Csort:date/gensim/BtN7uB1vgpc/tlvkLXqzJwAJ – outlines why this form of error-divvying makes sense for the average-forward, divide-backward case.

**Comment by spinkney** [comment]
This paper suggests that the original implementation and, subsequently, gensim's is wrong. https://arxiv.org/pdf/2012.15332.pdf

**Comment by piskvorky** [comment]
That is interesting, thanks for the link @spinkney !

**Comment by spinkney** [comment]
Yea, maybe, "wrong" is not the right word. As it clearly works, but not as well.

**Comment by Witiko** [comment]
@piskvorky @spinkney The paper makes some dubious claims:

> However, the CBOW negative sampling implementations in `word2vec.c` and Gensim incorrectly update each context vector by Eq. (1), without normalizing by the number of context words.

Gensim [does normalize][1] the gradient by the number of averaged input word and n-gram vectors.

> [In Gensim,] normalization by number of context words is only done for the ‘sum variant’ of CBOW.

That's not [what I see in the code][1]: Gensim does the right thing when `cbow_mean=1`, which is the default.  
In the sum variant (`cbow_mean=0`), we correctly do not normalize, because the loss function L is different.

 [1]: https://github.com/RaRe-Technologies/gensim/blob/master/gensim/models/word2vec_inner.pyx#L455

**Comment by Witiko** [comment]
@piskvorky

> Any appetite for trying this out in Gensim & benchmarking?   
> [...]   
> I'm less interested in what is "right / wrong" and more in what works better.

I am not convinced by the accuracy part of the paper due to the issues I discuss above, although I will admit I did not read the paper in depth. However, Figure 1 confirms my observations: Gensim's any2vec (at least prior to 4.0.0) does not scale well with the number of workers. That is definitely something that should be tackled, although I can't volunteer time to this at the moment.

![use1543](https://user-images.githubusercontent.com/603082/103484212-382d1a80-4ded-11eb-9dde-a2d13c92ac99.png)

**Comment by gojomo** [comment]
I've given the paper & their code a quick look, after seeing it on HN. You can see some of my comments [there](https://news.ycombinator.com/item?id=25611252). 

I'm not yet convinced their analysis is correct or their change offers any net benefits. Their own benchmarks show mixed results, with small differences either way. It'd be interesting if @tmikolov had any comments on their interpretation. 

I'm not sure their method of testing Gensim always with an `0.025` starting-`alpha`, while theirs with an `0.075` starting `alpha`, is fair/reasonable. (Instead of picking one value based on 'average' performance over many tasks, as long as they were grid-searching many values, why not let Gensim-at-its-best-`alpha`-each-task compare vs Koan-at-its-best-`alpha`-each-task?) 

Given that they observed that the (seldom-used & I believe even removed from `word2vec.c`) `cbow_mean=0` mode behaves more like their preference, I'm also surprised they didn't try benchmarking that mode (which also implies a different optimal `alpha`) against their variant. I have a hunch that any marginal benefits on any evaluation from their approach could be replicated in the 'standard' interpretation with some combo of alternate-alpha/`cbow_mean=0`/etc. B

Certainly, a `koan_cbow` option in Gensim would be easy to add, if it's useful for testing this variant or getting other benefits. 

Looking at their code, it also appears they've completely eliminated the `reduced_window`-method for stochastic weighting of neighboring words (which is used by `word2vec.c`, Gensim, & Facebook FastText. I'm not sure the motivation, but it could be a contributor to their different (sometimes-better, sometimes-worse) results on various benchmarks. In addition to options like @Witiko's learned/position-variable weighting, it'd be easy to add a flag to turn off the usual `reduced_windows` - allowing testing of that variant independently. (Sometimes people want "all tokens in the context window with no weighting", because their underlying data source doesn't have a true ordering - and I usually just recommend something like `window=1000000` in such cases, as it means that nearly-always, the effective-window will be the full text. But this hypothetical new flag could help that situation, & others, instead.)

Their use of an 'alias method' for negative-sampling looks like a separate, functionally-neutral but performance-enhancing change. It could be a nice separate optimization. I don't fully understand it yet – I think it'd use more memory than our current 'binary-search-over-list-of-split-points' approach, but remains O(1) (ignoring cache-locality effects) for arbitrarily large vocabularies. (The original `word2vec.c` approach of pre-filling a giant array was O(1) at a massive cost in memory that then also degraded its real performance due to cache misses, vs the binary-search that I replaced it with a long while back.) 

(There are likely other still low-hanging-fruit performance optimizations around some of the negative-sampling, reduced-windows, & frequent-word-downsampling code - such as using a faster RNG or simply reusing a smaller number of random-draws repeatedly. EG: just permute the values {1..window} once per epoch, then cycle through them for each center-word - it's not **as** random, but probably good enough. Something similar might work for negative-sampling or downsampling.)

Their graph showing Gensim performance gains for more threads in SG but not CBOW doesn't quite match my expectations - I thought they both petered out in gains (& started getting worse) past about a dozen threads, *unless* using `corpus_file` mode. It's unclear if they're using `corpus_file` mode in their tests - if so, their CBOW benchmarks are a surprise, if not the SG results are a surprise. The stark difference suggests there might be a simple fix in the CBOW code.

**Comment by gojomo** [comment]
A possibly-updated (later 2021?) version of the "Corrected CBOW..." paper is at: https://aclanthology.org/2021.insights-1.1.pdf. After another quick read, I remain unconvinced that this is either a 'correction' or even a 'consistent improvement'.

Firstly, as far as I'm concerned, the Mikolov (et al) implementations in `word2vec.c` & later Fasttext *are* the 'correct' implementation of the algorithms as originally specified, definitionally. (To the extent the papers had any unspecified versions, as a contemporaneous reference release, the Mikolov code is authoritative to resolve.) Other improvements/optimizations are of course always possible, but calling the original choice 'incorrect' or 'faulty' seems misguided. 

As above & per the tables in the paper, note that their results still show Gensim performing better than their approach on many of their evaluations!

Their improvements on *some* evaluations are pretty small, & (as is sadly rampant in papers) it's not crystal-clear that they've given comparison preexisting implementations (in this case Gensim) the same breadth of meta-parameter optimizations as their method. In particular, before claiming their approach is better, they should be sure to search the already-supported parameter ranges of Gensim for values that may closely approximate their changes, in particular:

1. changing the apportionment of updates across CBOW words has an effect very similar to a different `alpha` schedule, so the proper test is "best alpha in new implementation" vs "best alpha in reference implementation" (which might be higher). Note original `word2vec.c` would default to a *higher* `alpha` in CBOW mode. 
2. Testing both `cbow_mean` & sum modes, as supported by Gensim, which affect the same updates.
3. Testing both values for the new `shrink_windows` option (#3169), which matches a difference between their code & Mikolov-style window-treatment that's evident in their codebase (but unmentioned in their paper) that might account for changed end-results.

Their speed claims are also clouded by the inclusion of another, independent optimization of the negative-sampling which doesn't change the distribution. 

As per my prior comment, it'd be easy enough to prepare a patch with an optional `cbow_koan` mode, that narrowly changes the gradient calculation according to their main idea, then test that extensively to see where (if anywhere) it uniquely outperforms other parameter-mixes.
